docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,408 @@
+  name: qos-external-bar
+  namespace: default
+spec:
+  egress:
+  - dscp: 30
+    classifier:
+      to:
+      - cidr: 0.0.0.0/0
+```
+the equivalent of:
+```bash
+ovn-nbctl qos-add node1 to-lport 1000 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32 && tcp && tcp.dst == 22" dscp=59
+ovn-nbctl qos-add node1 to-lport 999 "ip4.src == <default_ns-qos-external-foo address set> && ip4.dst == 1.2.3.4/32" dscp=46
+ovn-nbctl qos-add node1 to-lport 998 "ip4.src == <default_ns address set> && ip4.dst == 0.0.0.0/0" dscp=30
+```
+will be executed.
+The podSelector in `qos-external-foo` NetworkQoS object is applied for both first and second NetworkQoS rules. So creating a new Pod that matches this selector in the namespace results in its IPs being added to corresponding Address Set.
+The `qos-external-bar` doesn't have podSelector, so NetworkQoS rule specified in this object is applied to all pods in the `default` namespace.
+
+```yaml
+kind: NetworkQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-storage-traffic
+  namespace: games
+spec:
+  podSelector:
+    matchLabels:
+      app: storage-client
+  egress:
+  - priority: 40
+    dscp: 59
+    bandwidth:
+      burst: 100
+      rate: 20000
+    classifier:
+      ports:
+        protocol: "tcp"
+        port: 471
+      to:
+      - podSelector:
+          matchLabels:
+            app: storage-target
+          namespaceSelector:
+            matchLabels:
+            name: storage-target
+---
+kind: NetworkQoS
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: qos-stream-traffic
+  namespace: games
+spec:
+  podSelector:
+    matchLabels:
+      app: stream-client
+  egress:
+  - priority: 40
+    dscp: 33
+    classifier:
+      ports:
+        protocol: "udp"
+        port: 399
+      to:
+      - podSelector:
+          matchLabels:
+          app: stream-client
+        namespaceSelector:
+          matchLabels:
+            name: stream-target
+```
+In the above example, The tcp traffic with dst port 471 from pods with label `app=storage-client` in `games` namespace will have dscp set to 59 and bandwidth limited to specified burst/rate towards pods having label `app=storage-target` residing on the namespace with label `name=storage-target`.
+The udp traffic with dst port 399 from pods with label `app=stream-client` in `games` namespace will have dscp set to 33 towards pods with label `app=stream-client` residing on the namespace with label `name=stream-target`.
+
+In addition it'll watch nodes to decide if further updates are needed, for example:
+when another node `node2` joins the cluster, the controller will attach the existing `NetworkQoS` object to its node local switch.
+
+IPv6 will also be supported, given the following `NetworkQoS`:
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: NetworkQoS
+metadata:
+  name: default
+  namespace: default
+spec:
+  egress:
+  - dscp: 48
+    classifier:
+      to:
+      - cidr: 2001:0db8:85a3:0000:0000:8a2e:0370:7330/124
```

Comment:
According to the documentation in line above, https://github.com/ovn-org/ovn-kubernetes/pull/4366/commits/45f83cb9b0dae01510dafad91bd4bfe78a6981c1#r1669343352
This should include the `ipBlock`, right?

```yaml
spec:
  egress:
  - dscp: 48
    classifier:
      to:
      - ipBlock:
            cidr: 2001:0db8:85a3:0000:0000:8a2e:0370:7330/124
```

Please check the other examples as well.

Reply:
yes, correct. this must be inside `ipBlock`, updated the examples.